contrib/check-for-api-changes.sh

```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+#
+# Checks the public API of crates, exits with non-zero if there are currently
+# changes to the public API not already committed to in the various api/*.txt
+# files.
+
+set -e
+
+export RUSTDOCFLAGS='-A rustdoc::broken-intra-doc-links'
+REPO_DIR=$(git rev-parse --show-toplevel)
+API_DIR="$REPO_DIR/api"
+CMD="cargo +nightly public-api --simplified"
+
+# cargo public-api uses nightly so the toolchain must be available.
+if ! cargo +nightly --version > /dev/null; then
+    echo "script requires a nightly toolchain to be installed (possibly >= nightly-2023-05-24)" >&2
+    exit 1
+fi
+
+pushd "$REPO_DIR" > /dev/null
+$CMD --no-default-features | sort --unique > "$API_DIR/no-default-features.txt"
+$CMD | sort --unique > "$API_DIR/default-features.txt"
+$CMD --no-default-features --features=alloc | sort --unique > "$API_DIR/alloc.txt"
+$CMD --no-default-features --features=core2 | sort --unique > "$API_DIR/core2.txt"
```

Comment:
Could you try sorting with `-n`? It would read better if the array types were sorted naturally.

Reply:
Using `-n` (or `-h`) causes the script to produce no output.

Reply:
WTF, that's some weird interaction between `-n` and `--unique`. You can fix it by writing `sort -n | uniq`

Reply:
So I think we have pushed bash as far as it can reasonably go, bash can't easily sort the following 4 lines as we want
```
impl<'a> hex_conservative::display::DisplayHex for &'a [u8; 9]
impl<'a> hex_conservative::display::DisplayHex for &'a [u8; 8]
pub type [u8; 128]::Err = hex_conservative::HexToArrayError
pub type [u8; 10]::Err = hex_conservative::HexToArrayError
```

Are you guys ok with a perl script to get the order to be:
```
impl<'a> hex_conservative::display::DisplayHex for &'a [u8; 8]
impl<'a> hex_conservative::display::DisplayHex for &'a [u8; 9]
pub type [u8; 10]::Err = hex_conservative::HexToArrayError
pub type [u8; 128]::Err = hex_conservative::HexToArrayError
```

Is there a better tool than perl for this? (FTR neither my perl nor awk are good enough to quickly write this but since we are going to use this script everywhere I can allocate some time to it.)

Reply:
Either awk or perl is fine by me. I can read, but not really write, those languages.

Reply:
Not a fan of perl but if it's the simplest choice for you I could tolerate it.